GenderPayGap.WebJob/Functions/Functions.Purge.cs

```diff
@@ -27,48 +28,71 @@ public async Task PurgeUsers([TimerTrigger("40 3 * * *" /* 03:40 once per day */
-                    .ToListAsync();
-                DateTime pinExpireyDate = VirtualDateTime.Now.AddDays(0 - Global.PinInPostExpiryDays);
-                foreach (User user in users)
+                //Ignore if they have verified PIN
+                if (user.UserOrganisations.Any(
+                    uo => uo.PINConfirmedDate != null || (uo.PINSentDate != null && uo.PINSentDate < pinExpiryDate)))
                 {
-                    //Ignore if they have verified PIN
-                    if (user.UserOrganisations.Any(
-                        uo => uo.PINConfirmedDate != null || uo.PINSentDate != null && uo.PINSentDate < pinExpireyDate))
-                    {
-                        continue;
-                    }
-
-                    var logItem = new ManualChangeLogModel(
-                        nameof(PurgeUsers),
-                        ManualActions.Delete,
-                        AppDomain.CurrentDomain.FriendlyName,
-                        nameof(user.UserId),
-                        user.UserId.ToString(),
-                        null,
-                        JsonConvert.SerializeObject(new {user.UserId, user.EmailAddress, user.JobTitle, user.Fullname}),
-                        null);
-                    _DataRepository.Delete(user);
-                    await _DataRepository.SaveChangesAsync();
-                    await Global.ManualChangeLog.WriteAsync(logItem);
+                    return;
                 }
 
-                LogFunctionEnd(runId, nameof(PurgeUsers), startTime);
+                var logItem = new ManualChangeLogModel(
+                    nameof(PurgeUsers),
+                    ManualActions.Delete,
+                    AppDomain.CurrentDomain.FriendlyName,
+                    nameof(user.UserId),
+                    user.UserId.ToString(),
+                    null,
+                    JsonConvert.SerializeObject(new {user.UserId, user.EmailAddress, user.JobTitle, user.Fullname}),
+                    null);
+            
+                DeleteUserAndAuditLogs(user);
+                await Global.ManualChangeLog.WriteAsync(logItem);
             }
             catch (Exception ex)
             {
                 LogFunctionError(runId, nameof(PurgeUsers), startTime, ex );
+            }
 
-                //Send Email to GEO reporting errors
-                await _Messenger.SendGeoMessageAsync("GPG - WEBJOBS ERROR", 
-                    $"Failed webjob ({nameof(PurgeUsers)}):{ex.Message}:{ex.GetDetailsText()}");
-                //Rethrow the error
-                throw;
+        }
+
+        private static void DeleteUserAndAuditLogs(User user)
+        {
+            var dataRepository = Program.ContainerIOC.Resolve<IDataRepository>();
+            
+            List<AuditLog> auditLogs = dataRepository
+                .GetAll<AuditLog>()
+                .Where(log => Equals(log.ImpersonatedUser, user))

```

Comment:
I think we should be comparing IDs here rather than the objects themselves.

The objects will be different instances (because they were created by different DataRepositories) so a normal `==` won't work.
You've used `Equals()` which routes through to the `User.Equals()` method (`GenderPayGap.Database.Models\Extensions\User.cs` line 214) which compares IDs.

I think `.Where(log => log.ImpersonatedUser.Id == user.Id)` would be clearest, since we only care about the IDs.

Reply:
Done